testcrate/tests/misc.rs

```diff
@@ -65,31 +65,46 @@ fn fuzz_values() {
 
 #[test]
 fn leading_zeros() {
-    use compiler_builtins::int::__clzsi2;
-    use compiler_builtins::int::leading_zeros::{
-        usize_leading_zeros_default, usize_leading_zeros_riscv,
-    };
-    fuzz(N, |x: usize| {
+    use compiler_builtins::int::leading_zeros::__clzsi2;
+    fuzz(N, |x: u32| {
+        if x == 0 {
+            return; // undefined value for an intrinsic
+        }
         let lz = x.leading_zeros() as usize;
         let lz0 = __clzsi2(x);
-        let lz1 = usize_leading_zeros_default(x);
-        let lz2 = usize_leading_zeros_riscv(x);
         if lz0 != lz {
             panic!("__clzsi2({}): std: {}, builtins: {}", x, lz, lz0);
         }
-        if lz1 != lz {
-            panic!(
-                "usize_leading_zeros_default({}): std: {}, builtins: {}",
-                x, lz, lz1
-            );
-        }
-        if lz2 != lz {
-            panic!(
-                "usize_leading_zeros_riscv({}): std: {}, builtins: {}",
-                x, lz, lz2
-            );
-        }
-    })
+    });
+
+    #[cfg(any(target_pointer_width = "32", target_pointer_width = "64"))]
+    {
+        use compiler_builtins::int::leading_zeros::__clzdi2;
+        fuzz(N, |x: u64| {
+            if x == 0 {
+                return; // undefined value for an intrinsic
+            }
+            let lz = x.leading_zeros() as usize;
+            let lz0 = __clzdi2(x);
+            if lz0 != lz {
+                panic!("__clzdi2({}): std: {}, builtins: {}", x, lz, lz0);
+            }
+        });
+    }
+    #[cfg(target_pointer_width = "64")]
+    {
+        use compiler_builtins::int::leading_zeros::__clzti2;
+        fuzz(N, |x: u128| {
+            if x == 0 {
+                return; // undefined value for an intrinsic
+            }
+            let lz = x.leading_zeros() as usize;
+            let lz0 = __clzti2(x);
+            if lz0 != lz {
+                panic!("__clzti2({}): std: {}, builtins: {}", x, lz, lz0);
+            }
+        });
+    }
```

Comment:
What exactly is happening on 16 & 32-bit platforms for the gating - is LLVM lowering `.leading_zeros()` to a symbol that is missing? From CI it looks like MSVC is missing `___clzti2` on arm and 32-bit so you'll have to update the gates.

If this logic is a bit more complex than just pointer sizes, it might be better to add a `NoSysU128Clz` option to https://github.com/rust-lang/compiler-builtins/blob/06db2de1c098f9c3cf69a52e7a273f19f0dce061/testcrate/build.rs#L5-L10, update that file, and then gate on that (easier if the gates need to be reused).

Reply:
I gated these, and the intrinsics themselves, the same way the compilers do (well, libgcc does so they have to follow along?). Makes no sense to test for intrinsics which aren't ever used.

As for 128 bit on 64-bit msvc - I guess they don't have 128 bit int somehow, according to compiler-rt. I just took "optimized c" attribute off the implementation instead of trying to gate it. Single branch isn't that bad.

Reply:
Ah, I missed that the gates were on the intrinsics as well. Could you please remove them? The reasons that `libgcc` doesn't supply a specific intrinsic on specific platforms may vary (no integers of that type, GCC itself doesn't use them, platform bugs, untested, just not done yet etc). But if we don't have any specific reason not to make them available, we should just provide them. And if there is a specific reason not to provide them (bugs), it needs to be commented.

Also a lot of the purpose of this library is to fill in for the shortcomings of libgcc, so it doesn't make any sense to copy their limitations :).

> well, libgcc does so they have to follow along?

Unfortunately this isn't very accurate, LLVM (and GCC) both pretty frequently emit symbols that aren't available on lesser-known platforms. Rust hits this more often because we make less common types available on all platforms (e.g. `u128`, `f16`, `f128`) where there might not be a C equivalent available.

Not like this specific symbol is likely to be a problem anywhere. But we should avoid unnecessary `cfg` because it just means somebody might have to chase them down in the future and wonder why it's disabled.

Reply:
Removed all the gate. Let the flood begin!

As for libgcc and their builtins, the reason there is simple: that's how they do it. Basically they have two and only two functions __clzSI2 and __clzDI2 (note uppercase letters). Despite their names, they operate on native word size and double that respectively. These function names (and argument types) get macroed into whatever proper name and argument type is for the platform.

Same is true for many other builtins there.

Reply:
ping?

Reply:
Thanks for the change. I think you might be able to remove some C sources that we are building (mentioned below) but otherwise the implementation looks good to me now.